### PR TITLE
Port SOCKS5 authentication from muon.

### PR DESCRIPTION
**Pull request by riastradh-brave**
This ports the patches, makes them work, and adds unit tests, in separate commits.  (If the unit tests should go again, they can be removed -- much easier this time since all the patches are in different files.)

fix https://github.com/brave/brave-browser/issues/358

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [x] Request a security/privacy review as needed.

## Test Plan:
After a build, say a Debug build:
- Build and run the unit tests:
  1. `ninja -C src/out/Debug net:net_unittests`
  2. `./src/out/Debug/net_unittests --gtest_filter='ProxyServerTest.*:SOCKS5ClientSocketTest.*'`
- Manually confirm the behaviour with Tor:
  1. Launch a tor process listening on (say) 9250, e.g. by starting browser-laptop and opening a private tab with Tor.
  2. Launch two independent Brave instances with different proxy server usernames and passwords, e.g.:
     - `./src/out/Debug/brave --user-data-dir=/tmp/test1 --proxy-server=socks5://foo:bar@127.0.0.1:9250 &`
     - `./src/out/Debug/brave --user-data-dir=/tmp/test2 --proxy-server=socks5://baz:quux@127.0.0.1:9250 &`
  3. Load https://check.torproject.org in the two browsers and confirm that they have different IP addresses.

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions

### CODE REVIEW

File: patches/net-base-host_port_pair.cc.patch [hunk 1]
```diff
@@ -0,0 +1,31 @@
+diff --git a/net/base/host_port_pair.cc b/net/base/host_port_pair.cc
+index ec8248449dbcd4156e361fcc8781ce0712f26a02..77f5a3653083625f8043147d42297dba05bb79a9 100644
+--- a/net/base/host_port_pair.cc
++++ b/net/base/host_port_pair.cc
+@@ -18,9 +18,11 @@
+ namespace net {
+ 
+ HostPortPair::HostPortPair() : port_(0) {}
++#if 0                           // brave override
```
Comment (bridiver) [inline patches/net-base-host_port_pair.cc.patch]: these should use BRAVE_CHROMIUM_BUILD

File: chromium_src/net/socket/socks5_client_socket_auth.cc [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+#include "net/base/io_buffer.h"
+#include "net/socket/socks5_client_socket_auth.h"
+
+namespace net {
+
+SOCKS5ClientSocketAuth::SOCKS5ClientSocketAuth(
+    std::unique_ptr<ClientSocketHandle> transport_socket,
+    const HostResolver::RequestInfo& req_info,
+    const NetworkTrafficAnnotationTag& traffic_annotation,
+    const HostPortPair& proxy_host_port)
```
Comment (bridiver) [inline chromium_src/net/socket/socks5_client_socket_auth.cc]: I'm sure we talked about doing a subclass/friend for HostPortPair, but I can't remember why we didn't. The only place we create the auth version is in our own code so it seems like it should work?

File: patches/net-BUILD.gn.patch [hunk 3]
```diff
@@ -0,0 +1,14 @@
+diff --git a/net/BUILD.gn b/net/BUILD.gn
+index 9d59189d80aea8f99b73020f36a578a96667af37..d5aa98b8f2f44f6acec58c124c262bd1ddcabe31 100644
+--- a/net/BUILD.gn
++++ b/net/BUILD.gn
+@@ -355,6 +355,9 @@ component("net") {
+     "ssl/token_binding.h",
+     "third_party/quic/core/quic_error_codes.cc",
+     "third_party/quic/core/quic_error_codes.h",
++    "//brave/chromium_src/net/base/host_port_pair_auth.cc",
```
Comment (bridiver) [inline patches/net-BUILD.gn.patch]: we can actually add these using a patch on the host_port_pair cc file

File: patches/net-BUILD.gn.patch [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
++    "//brave/chromium_src/net/base/url_auth_util.cc",
```
Comment (bridiver) [inline patches/net-BUILD.gn.patch]: same thing applies to these

File: patches/net-base-host_port_pair.cc.patch (hunk 1 above)
Comment (riastradh-brave) [reply patches/net-base-host_port_pair.cc.patch]: 5e8611a27cd370eb31c4b0e911a38e9fc0341c9f

File: chromium_src/net/socket/socks5_client_socket_auth.cc (hunk 2 above)
Comment (riastradh-brave) [reply chromium_src/net/socket/socks5_client_socket_auth.cc]: Would presumably require making several of HostPortPair's member functions virtual.  Not sure what the tradeoff is of the extra members vs making those functions virtual is.  Originally I didn't touch HostPortPair at all, but rather put it into ProxyServer.

File: patches/net-BUILD.gn.patch (hunk 3 above)
Comment (riastradh-brave) [reply patches/net-BUILD.gn.patch]: cb908db81cf50cf6eafbdbed8ffe7cd683188010

File: patches/net-BUILD.gn.patch (hunk 4 above)
Comment (riastradh-brave) [reply patches/net-BUILD.gn.patch]: 614d17f6149feded4eb20f9049a9050ba1a5cfd2
a9a045e0dd404ef76cc00a727f41118cc4ecc17d

**Comment by riastradh-brave** [comment]
I eliminated the BUILD.gn and net_log_event_type_list.h patches using overrides.

I could go a little further with overrides, but in some cases I'm reluctant to do that because if the code I'm overriding has changed enough, I _want_ there to be a merge conflict that requires human attention.

I could replace the definition of ParseHostAndPort, but there are various callers beyond proxy_server.cc.  I think it is safer to define a new function with the semantics we need there and patch proxy_server.cc to use it -- again, if the caller in that .cc file changes, it probably deserves human attention.

I could reduce the diff in host_port_pair.h a tiny bit by using a subclass for the version with an authority, but that would require making a number of member functions virtual, including several that are written in-line in host_port_pair.h, which strikes me as a net negative.

**Comment by bridiver** [comment]
I spent some time going through this with @riastradh-brave and unfortunately we couldn't use the preprocessor method to subclass HostPortPair because it gets passed by value (copy constructor)

**Pull request merged**